control-plane-operator/controllers/hostedcontrolplane/dnsoperator/dnsoperator.go

```diff
@@ -0,0 +1,190 @@
+package dnsoperator
+
+import (
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/kas"
+	"github.com/openshift/hypershift/support/config"
+	"github.com/openshift/hypershift/support/util"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilpointer "k8s.io/utils/pointer"
+)
+
+const (
+	// dnsOperatorContainerName is the name of the operator container.
+	dnsOperatorContainerName = "dns-operator"
+)
+
+// Images stores the image pullspecs for the images that the DNS operator
+// deployment references.
+type Images struct {
+	DNSOperator   string
+	CoreDNS       string
+	KubeRBACProxy string
+	CLI           string
+}
+
+// Params stores the parameters that are required to configure the DNS operator
+// deployment on HyperShift, as well as some additional information about the
+// deployment.
+type Params struct {
+	// ReleaseVersion is the HyperShift release version.
+	ReleaseVersion string
+	// AvailabilityProberImage is the image for the prober, which runs on
+	// the management cluster and probes the DNS operator.
+	AvailabilityProberImage string
+	// Images has the image pullspecs that the DNS operator deployment
+	// references.
+	Images Images
+	// OwnerRef is a reference to the HyperShift control plane.
+	OwnerRef config.OwnerRef
+	// DeploymentConfig has information about the DNS operator deployment.
+	DeploymentConfig config.DeploymentConfig
+}
+
+// NewParams creates a new Params object for a DNS operator deployment.
+func NewParams(hcp *hyperv1.HostedControlPlane, version string, images map[string]string) Params {
+	p := Params{
+		Images: Images{
+			DNSOperator:   images["cluster-dns-operator"],
+			CoreDNS:       images["coredns"],
+			KubeRBACProxy: images["kube-rbac-proxy"],
+			CLI:           images["cli"],
+		},
+		ReleaseVersion:          version,
+		AvailabilityProberImage: images[util.AvailabilityProberImageName],
+		OwnerRef:                config.OwnerRefFrom(hcp),
+	}
+
+	p.DeploymentConfig.AdditionalAnnotations = map[string]string{
+		"target.workload.openshift.io/management": `{"effect": "PreferredDuringScheduling"}`,
+	}
+	p.DeploymentConfig.AdditionalLabels = map[string]string{
+		"name": "dns-operator",
```

Comment:
this should probably include label ControlPlaneComponent = "hypershift.openshift.io/control-plane-component"